src/rgw/rgw_gc.cc

```diff
@@ -314,10 +314,10 @@ class RGWGCIOManager {
 public:
   RGWGCIOManager(const DoutPrefixProvider* _dpp, CephContext *_cct, RGWGC *_gc) : dpp(_dpp),
                                                   cct(_cct),
-                                                  gc(_gc),
-                                                  remove_tags(cct->_conf->rgw_gc_max_objs),
-                                                  tag_io_size(cct->_conf->rgw_gc_max_objs) {
+                                                  gc(_gc) {
     max_aio = cct->_conf->rgw_gc_max_concurrent_io;
+    remove_tags.reserve(min(static_cast<int>(cct->_conf->rgw_gc_max_objs), rgw_shards_max()));
```

Comment:
how does this enforce limit on rgw_gc_max_objs?

Reply:
it will not allow to process over `rgw_shards_max()` value. Without this change, if we set rgw_gc_max_objs for 70000, the RGWGC::remove tries to `gc_aio_operate` operate on not existent objects

Reply:
this reserves space for thus many elements, right?  is that needed because we're using emplace?  do we really want to reserve the max, though?

Reply:
i.e., would it make sense to enforce a max -size-, and reject an attempt to reshard to a larger size?  separately, maybe we could avoid the pre-sizing optimization, or else adjust things around it to avoid expecting the current max size?

Reply:
Looking at the code, we're currently allocating "unlimited" (according to configuration option) `remove_tags` and `tag_io_size`. This reserve won't allow to allocate more than RGW_SHARDS_PRIME_1. We have a lot of static references to these vectors, so yes I'm agree that RGWGCIOManager should be refactored, but this require more work. As for now I think this fix will block, deletion of objects over v RGW_SHARDS_PRIME_1.

Reply:
i don't understand this change to use `reserve()` instead of constructing the vector with a given size. this leaves the vector's `size()` at 0, meaning that:
* all attempts to index it will be out of bounds:
```c++
auto& rt = remove_tags[index];
```
* for-loops will iterate over an empty vector:
```c++
  void flush_remove_tags() {
    int index = 0;
    for (auto& rt : remove_tags) {
```

instead, i would just default-construct the empty vectors, and wait to resize them until `initialize()`:
```diff
 void RGWGC::initialize(CephContext *_cct, RGWRados *_store) {
   cct = _cct;
   store = _store;
 
   max_objs = min(static_cast<int>(cct->_conf->rgw_gc_max_objs), rgw_shards_max());
+  remove_tags.resize(max_objs);
+  tag_io_size.resize(max_objs);
```

Reply:
@cbodley thank you

Reply:
@cbodley sure, `resize` is better, I placed it under constructor of IOManager, because the `RGWGC::initialize` doesn't know about it